proposed-api.md

```diff
@@ -0,0 +1,213 @@
+
+For the purpose of the discussion below, consider the following VPC configurations:
+
+- vpc1, has all addresses 10.1.0.0/16
+- vpc2, has all addresses 10.2.0.0/16
+- vpc3, has all addresses 10.3.0.0/16
+
+- external1 is connected to the internet and peered via BGP to provider A
+- external2 is connected to the internet and peered via BGP to provider B
+
+## Expose
+
+To expose a set of IP addresses for use by other VPCs and externals, an Expose object is used.  The gits of an expose object is as follows:
+```
+kind: gateway/expose
+metadata:
+  name: expose-10-1-1-0--24
+spec:
+  # When traffic is received at the exposed IPs we need to be able
+  # to return traffic back.  This creates an implied route.
+  # The default choice here should be "stateful" where we use a
+  # flow table like a stateful firewall to return the exact traffic
+  # to the host that generated the original flow, but this is effectively
+  # flow-based routing which may not scale, so we should allow other options.
+  # 
+  # Valid values are: static, stateful
+  #
+  return-route-type: static
+  expose:
+    ip: 10.1.1.0/24 # These ips must belong to the VPC
+    as: 192.168.1.0/24 # This implies static DNAT from these IPs to ip
+    map-ports: # Support for PAT on dest
+      - external: 80
+        internal: 8080
+  # Additional filter that will only allow connections from vpc2 and vpc3
+  # but not other vpcs or externals.  We probably need a way to wildcard
+  # vpcs or use a regexp or something to make it easier to specify sources.
+  #
+  # This does *not* imply any route though, this is just a filter as the
+  # name suggests.  It will drop packets with other origins.
+  filter:
+    allow:
+      - from: { vpc: vpc2 }
+        to: { port: 80 }
+      - from-vpc: { vpc: vpc2 }
+        to: { port: 80 }
+```
+
+Note that an Expose object does not by itself create an routes in the gateway 
+to allow traffic from any source to any destination, it just tells the gateway
+that traffic from outside the VPC to these addresses should be allowed.
+
+## Route
+
+`Route` objects do roughly what the name implies, they create routes in the
+gateway to allow traffic to flow from a source VPC to a set of IP addresses 
+made visible by an `Expose` object.  
+
+Because they are routes, the lookup is based only on destination IP in the VRF for the source VPC.  
+```
+# This route has no effect unless there is an Expose object with
+# these IP addresses.
+#
+kind: gateway/route
+metadata:
+  name: web-traffic-vpc2-to-vpc1
+spec:
+  # Note, longer term we want this whole tuple to specifiy a VRF
+  # for now, we'll live with just the vpc
+  src: 
```

Comment:
Would be great to dedup it with Expose object data little bit. Can we just make src reference the Expose as well?